controllers/resource_reconciler.go

```diff
@@ -261,3 +303,83 @@ func readyStatus(ctx context.Context, u *unstructured.Unstructured) (bool, error
 	// no ready condition is set
 	return false, nil
 }
+
+// partitionResources splits the sets of resources in spec and the current set
+// of owned, existing resources into three groups:
+// - unobservedTypeResources are of a type not yet known to this owning CAPZ resource.
+// - observedTypeResources are of a type already known to this owning CAPZ resource.
+// - deletedResources exist but are not defined in spec.
+func partitionResources(
+	ownedKinds sets.Set[metav1.TypeMeta],
+	specs []*unstructured.Unstructured,
+	ownedObjs []*metav1.PartialObjectMetadata,
+) (
+	unobservedTypeResources []*unstructured.Unstructured,
+	observedTypeResources []*unstructured.Unstructured,
+	deletedResources []*metav1.PartialObjectMetadata,
+) {
+	for _, spec := range specs {
+		typeMeta := metav1.TypeMeta{
+			APIVersion: spec.GetAPIVersion(),
+			Kind:       spec.GetKind(),
+		}
+		if ownedKinds.Has(typeMeta) {
+			observedTypeResources = append(observedTypeResources, spec)
+		} else {
+			unobservedTypeResources = append(unobservedTypeResources, spec)
+		}
+	}
+
+	for _, owned := range ownedObjs {
+		if !slices.ContainsFunc(specs, metadataRefersToResource(owned)) {
+			deletedResources = append(deletedResources, owned)
+		}
+	}
+	return
+}
+
+func statusResource(resource *unstructured.Unstructured) infrav1alpha.StatusResource {
+	gvk := resource.GroupVersionKind()
+	return infrav1alpha.StatusResource{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind,
+		Name:    resource.GetName(),
+	}
+}
+
+func metadataRefersToResource(metadata *metav1.PartialObjectMetadata) func(*unstructured.Unstructured) bool {
+	return func(u *unstructured.Unstructured) bool {
+		// Version is not a stable property of a particular resource. The API version of an ASO resource may
+		// change in the CAPZ spec from v1 to v2 but that still represents the same underlying resource.
+		return metadata.GroupVersionKind().GroupKind() == u.GroupVersionKind().GroupKind() &&
+			metadata.Name == u.GetName()
+	}
+}
+
+func parseOwnedKinds(value string) (sets.Set[metav1.TypeMeta], error) {
```

Comment:
We should probably have a unit test for this function validating a scenario with an ownerKind with multiple `/` in it.